library/psa_crypto_cipher.c

```diff
@@ -42,45 +42,97 @@ static psa_status_t mbedtls_cipher_validate_values(
     psa_algorithm_t alg,
     psa_key_type_t key_type)
 {
-    switch (alg) {
-        case PSA_ALG_STREAM_CIPHER:
-        case PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CHACHA20_POLY1305, 0):
-            if (key_type != PSA_KEY_TYPE_CHACHA20) {
-                return PSA_ERROR_NOT_SUPPORTED;
-            }
-            break;
+    /* Reduce code size - hinting to the compiler about what it can assume allows the compiler to
+       eliminate bits of the logic below. */
+#if !defined(PSA_WANT_KEY_TYPE_AES)
+    MBEDTLS_ASSUME(key_type != PSA_KEY_TYPE_AES);
+#endif
+#if !defined(PSA_WANT_KEY_TYPE_ARIA)
+    MBEDTLS_ASSUME(key_type != PSA_KEY_TYPE_ARIA);
+#endif
+#if !defined(PSA_WANT_KEY_TYPE_CAMELLIA)
+    MBEDTLS_ASSUME(key_type != PSA_KEY_TYPE_CAMELLIA);
+#endif
+#if !defined(PSA_WANT_KEY_TYPE_CHACHA20)
+    MBEDTLS_ASSUME(key_type != PSA_KEY_TYPE_CHACHA20);
+#endif
+#if !defined(PSA_WANT_KEY_TYPE_DES)
+    MBEDTLS_ASSUME(key_type != PSA_KEY_TYPE_DES);
+#endif
+#if !defined(PSA_WANT_ALG_CCM)
+    MBEDTLS_ASSUME(alg != PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CCM, 0));
+#endif
+#if !defined(PSA_WANT_ALG_GCM)
+    MBEDTLS_ASSUME(alg != PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_GCM, 0));
+#endif
+#if !defined(PSA_WANT_ALG_STREAM_CIPHER)
+    MBEDTLS_ASSUME(alg != PSA_ALG_STREAM_CIPHER);
+#endif
+#if !defined(PSA_WANT_ALG_CHACHA20_POLY1305)
+    MBEDTLS_ASSUME(alg != PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CHACHA20_POLY1305, 0));
+#endif
+#if !defined(PSA_WANT_ALG_CCM_STAR_NO_TAG)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CCM_STAR_NO_TAG);
+#endif
+#if !defined(PSA_WANT_ALG_CTR)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CTR);
+#endif
+#if !defined(PSA_WANT_ALG_CFB)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CFB);
+#endif
+#if !defined(PSA_WANT_ALG_OFB)
+    MBEDTLS_ASSUME(alg != PSA_ALG_OFB);
+#endif
+#if !defined(PSA_WANT_ALG_XTS)
+    MBEDTLS_ASSUME(alg != PSA_ALG_XTS);
+#endif
+#if !defined(PSA_WANT_ALG_ECB_NO_PADDING)
+    MBEDTLS_ASSUME(alg != PSA_ALG_ECB_NO_PADDING);
+#endif
+#if !defined(PSA_WANT_ALG_CBC_NO_PADDING)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CBC_NO_PADDING);
+#endif
+#if !defined(PSA_WANT_ALG_CBC_PKCS7)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CBC_PKCS7);
+#endif
+#if !defined(PSA_WANT_ALG_CMAC)
+    MBEDTLS_ASSUME(alg != PSA_ALG_CMAC);
+#endif
```

Comment:
This certainly looks like the most maintainable version - nice!

Reply:
From a readability perspective I quite like it, but what about compilers other than GCC and Clang? I think IAR for example is used in the embedded space - will it benefit from `MBEDTLS_ASSUME`?

Reply:
Would something like this work

```
#elif defined(__IAR_SYSTEMS_ICC__)
static inline __noreturn mbedtls_iar_unreachable(void)
{
    /* Will probably need to disable warning about __noreturn function returning */
    return;
}

#define MBEDTLS_ASSUME(cond) do { if (!(cond)) mbedtls_iar_unreachable(); } while (0)
```

Reply:
Yes, IAR 9.4 benefits (via the `__builtin_unreachable` option) by about the same amount

Reply:
I'm not familiar enough with IAR to judge, but I'll make a more general version of my point: by relying on compiler-specific extensions to achieve code size reduction, we run the risk of leaving out some compilers / make ourselves responsible for taking care of more compilers.

I'm not sure the code before `MBEDTLS_ASSUME` was bad enough in order to justify that new cost? (Probably depends how many other occurences we expect to have where `MBEDTLS_ASSUME` can improve the readability-code-size trade-off.)

Reply:
The risk is that niche compilers don't get code-size benefits (but no functional difference in behaviour). That's not a big problem; we can't optimise for code size with every compiler. Doing our best for gcc, clang, MSVC and IAR is probably a sensible place to be.

I think it's an acceptable trade-off for better readability - the ifdef's we had before were making the logic tricky to follow and IMO definitely increasing the risk of a bug.

Reply:
Also note that we didn't have to do anything special for IAR - it just picked up the `builtin_unreachable` option. I'd guess a good proportion of other compilers will do the same, I think it's fairly widely supported.